### ISSUE TITLE
Problem with DBServer:SetSelectiveRelation()

### ISSUE THREAD

**Issue by cpyrgas** [X-Sharp/XSharpPublic#525]
Try the following code with the attached (cut down) dbfs, see the comments, X# returns different results to those of VO (and Vulcan). This is what probably causes some issues in parent/detail browsers that I am investigating.

```
FUNCTION Start() AS VOID
LOCAL parent , child AS dbserver

parent := DbServer{"C:\Test\Kunden.dbf",TRUE,,"DBFCDX"}
child := DbServer{"C:\Test\Rechnung.dbf",TRUE,,"DBFCDX"}
child:SetOrder( 1 )

? parent:SetSelectiveRelation(child, #KDNO )
parent:GoTop()

? child:RecNo // 7

? "Skipping forward"
child:Skip(+1)
? child:RecNo, child:EoF // 7 TRUE, SHOULD BE: 12 TRUE
child:Skip(+1)
? child:RecNo, child:EoF // 7 TRUE, SHOULD BE: 12 TRUE
child:Skip(+1)
? child:RecNo, child:EoF // 7 TRUE, SHOULD BE: 12 TRUE

? "Skipping backward"
child:Skip(-1)
? child:RecNo, child:BoF, child:EoF // 9 F F, SHOULD BE: 7 F F
child:Skip(-1)
? child:RecNo, child:BoF, child:EoF // 8 F F, SHOULD BE: 7 T F
child:Skip(-1)
? child:RecNo, child:BoF, child:EoF // 7 F F, SHOULD BE: 7 T F
child:Skip(-1)
? child:RecNo, child:BoF, child:EoF // 7 T F, SHOULD BE: 7 T F

parent:Close()
child:Close()

```
[TestDbf.zip](https://github.com/X-Sharp/XSharpPublic/files/5549333/TestDbf.zip)

**Comment by cpyrgas** [comment]
Self contained VS project sample:

[ConsoleApplication1.zip](https://github.com/X-Sharp/XSharpPublic/files/5549371/ConsoleApplication1.zip)

**Comment by cpyrgas** [comment]
Self contained sample without dependency to DBServer:


```
FUNCTION Start() AS VOID
LOCAL cbExpression AS USUAL
LOCAL u1,u2 AS USUAL
LOCAL cPath AS STRING
LOCAL n AS INT
cPath := "C:\test\"

RddSetDefault("DBFCDX")

DbCreate(cPath + "parent" , {{"FLD" , "N" , 3 , 0}})
DbCreate(cPath + "child" , {{"FLD" , "N" , 3 , 0}})


DbUseArea(TRUE,"DBFCDX", cPath + "parent", "parent" , TRUE)
parent->DbAppend()
parent->FieldPut(1,3)

DbUseArea(TRUE,"DBFCDX", cPath + "child", "child" , TRUE)
child->DbCreateIndex(cPath + "child","FLD")
FOR n := 1 UPTO 5
child->DbAppend()
child->FieldPut(1,n)
NEXT

cbExpression := &( "{ || FLD }")
child->DbGoBottom()
? "evaluate in bottom, must be 5:", child->Eval( cbExpression ) // 3

? parent->VoDbSetRelation("child" , cbExpression ,"FLD")
parent->DbGoTop()

child->DbSkip(0) // this throws a runtime error in __FieldGet()

? "Recno must be 3:",child->RecNo() 
?
? "evaluate after SetRelation(), must be 3:", child->Eval( cbExpression )
? "Recno should be 3:", child->RecNo()

u1 := u2 := 3
? child->VoDbOrderInfo( DBOI_SCOPETOP	 , "", NIL, REF u1 )
? child->VoDbOrderInfo( DBOI_SCOPEBottom, "", NIL, REF u2 )
parent->DbGoTop()


child->DbGoTop()
? "recno after set scope, must be 3:", child->RecNo()

? "Skipping forward, all must be 6 , FALSE, TRUE"
child->DbSkip(+1)
? child->RecNo() , child->Bof() , child->Eof()
child->DbSkip(+1)
? child->RecNo() , child->Bof() , child->Eof()
child->DbSkip(+1)
? child->RecNo() , child->Bof() , child->Eof()

? "Skipping backward"
child->DbSkip(-1)
? child->RecNo() , child->Bof() , child->Eof() // must be 3, false, false
child->DbSkip(-1)
? child->RecNo() , child->Bof() , child->Eof() // must be 3, TRUE, false
child->DbSkip(-1)
? child->RecNo() , child->Bof() , child->Eof() // must be 3, TRUE, false
child->DbSkip(-1)
? child->RecNo() , child->Bof() , child->Eof() // must be 3, TRUE, false

DbCloseAll()

```

**Comment by RobertvanderHulst** [comment]
Note: I tested this with Vulcan (4) and VO (s.8 SP3) and there are also several errors in Vulcan and VO with this code.
They both report 
`recno after set scope, must be 3: 1`
and after that operation several of the other values are incorrect too.

**Comment by cpyrgas** [comment]
Oops, sorry about the mistakes in the tests! I had spent a lot of hours investigating this problem in the user's app, so was very tired when I logged it. Should had better waited for the next day.

About VO and Vulcan, I just checked again, I am getting the exact results I am mentioning in the code, including this "recno after set scope", it indeed returns 3 here in VO. Maybe you did some other change to the code? Can't think why you might be getting different results..
